### PR TITLE
8290732: JNI DestroyJavaVM can start shutdown when one non-daemon thread remains

### PR DESCRIPTION
**Pull request by dholmes-ora**
Please review this fix for a problem discovered by @stuart-marks in the course of examining the VM shutdown behaviour. The VM code assumed that only unattached threads called JNI's DestroyJavaVM and so they were always attached as non-daemon threads. But it is perfectly valid to call DestroyJavaVM from an already attached thread, which could be a daemon. The fix simply checks whether the caller is a daemon or not and adjusts the expected count of active threads to see. There is also an adjustment to the thread termination logic to also notify at the right time.

Thanks to @stuart-marks for the reproducer in JBS - the longest part of this by many hours was converting the test over the jtreg. :)

Testing:
 - the new test on all core platforms
 - tiers 1-3
Thanks.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8290732](https://bugs.openjdk.org/browse/JDK-8290732): JNI DestroyJavaVM can start shutdown when one non-daemon thread remains

### Reviewers
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)
 * [Robbin Ehn](https://openjdk.org/census#rehn) (@robehn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9803/head:pull/9803` \
`$ git checkout pull/9803`

Update a local copy of the PR: \
`$ git checkout pull/9803` \
`$ git pull https://git.openjdk.org/jdk pull/9803/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9803`

View PR using the GUI difftool: \
`$ git pr show -t 9803`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9803.diff">https://git.openjdk.org/jdk/pull/9803.diff</a>

</details>

### CODE REVIEW

File: test/hotspot/jtreg/runtime/jni/daemonDestroy/TestDaemonDestroy.java
Comment (tstuefe) [inline test/hotspot/jtreg/runtime/jni/daemonDestroy/TestDaemonDestroy.java:72]: Should we not add an `oa.shouldContain("T1 finished")` ?

File: test/hotspot/jtreg/runtime/jni/daemonDestroy/TestDaemonDestroy.java
Comment (tstuefe) [inline test/hotspot/jtreg/runtime/jni/daemonDestroy/TestDaemonDestroy.java:59]: Nit: better arg handling?
- If you want to just have one char arg, pass 'D' and 'N' for demon/nondemon mode instead of "1" and "".
- Instead of using args.length, maybe actually parse the input argument, or specify D and N there too and pass it on.

File: test/hotspot/jtreg/runtime/jni/daemonDestroy/TestDaemonDestroy.java
Comment (tstuefe) [inline test/hotspot/jtreg/runtime/jni/daemonDestroy/TestDaemonDestroy.java:36]: If you do two @test sections, you can give them speaking names(e.g. @test id=demon) and they can run in parallel.

**Review by tstuefe — COMMENTED** [review]
Hi David,

that was interesting. I always thought you have to detach before destroying the VM.

I looked into https://docs.oracle.com/javase/7/docs/technotes/guides/jni/spec/invocation.html and it says so:

"Detaching from the VM: A native thread attached to the VM must call DetachCurrentThread() to detach itself before exiting."

Cheers, Thomas

File: make/test/JtregNativeHotspot.gmk
Comment (magicus) [inline make/test/JtregNativeHotspot.gmk:869]: Actually this line was correctly indented, and all other incorrectly... If you can be bothered to fix it, please make sure there is just a two space indentation in the "if" block.

File: test/hotspot/jtreg/runtime/jni/daemonDestroy/Main.java
Comment (dholmes-ora) [reply test/hotspot/jtreg/runtime/jni/daemonDestroy/Main.java:36]: I will delete the comment. It was really a note-to-self after I finally debugged the NoSuchMethodError I was getting because I'd copied code that looks for a regular "main".

File: test/hotspot/jtreg/runtime/jni/daemonDestroy/TestDaemonDestroy.java
Comment (dholmes-ora) [reply test/hotspot/jtreg/runtime/jni/daemonDestroy/TestDaemonDestroy.java:59]: I reworked it slightly so the Java side doesn't pass "" and the native code interprets any additional arg as "use daemon".

File: test/hotspot/jtreg/runtime/jni/daemonDestroy/TestDaemonDestroy.java
Comment (dholmes-ora) [reply test/hotspot/jtreg/runtime/jni/daemonDestroy/TestDaemonDestroy.java:36]: I didn't think that was warranted/necessary here.

File: make/test/JtregNativeHotspot.gmk
Comment (dholmes-ora) [reply make/test/JtregNativeHotspot.gmk]: Fixed here and throughout the file.

**Comment by dholmes-ora** [comment]
Thanks for looking at this @tstuefe and @magicus

**Comment by stuart-marks** [comment]
Thanks for fixing this. I'm glad my reproducer gave you a good start, but you had to do all the grunt work of error checking and jtreg conversion. :-) Things look reasonable to me, but I'm not too familiar with this code and with Hotspot testing, so I'll defer to other reviewers.

File: test/hotspot/jtreg/runtime/jni/daemonDestroy/TestDaemonDestroy.java
Comment (tstuefe) [reply test/hotspot/jtreg/runtime/jni/daemonDestroy/TestDaemonDestroy.java:36]: I disagree. Splitting the test leads causes better runtime (on my 24 core machine: 10.5 seconds vs 16 seconds). It also makes the jtr file easier to read, since it only contains one test. It also is simpler to run a selected test if you want to reproduce an error.

It is regrettable that splitting the test section causes too much duplication. Maybe we could improve the jtreg syntax in the future.

File: test/hotspot/jtreg/runtime/jni/daemonDestroy/TestDaemonDestroy.java
Comment (dholmes-ora) [reply test/hotspot/jtreg/runtime/jni/daemonDestroy/TestDaemonDestroy.java:78]: I will add this too.

File: test/hotspot/jtreg/runtime/jni/daemonDestroy/TestDaemonDestroy.java
Comment (dholmes-ora) [reply test/hotspot/jtreg/runtime/jni/daemonDestroy/TestDaemonDestroy.java:36]: But you also have to duplicate the build instructions and building takes longer than execution on my systems (7 seconds versus 5) so if I had to pay for machine time I'd actually be worse off now!

**Review by tstuefe — APPROVED** [review]
Looks good.

File: test/hotspot/jtreg/runtime/jni/daemonDestroy/TestDaemonDestroy.java
Comment (tstuefe) [reply test/hotspot/jtreg/runtime/jni/daemonDestroy/TestDaemonDestroy.java:36]: Oh, you are right. That's unfortunate.

**Review by tstuefe — APPROVED** [review]
Still good.

**Comment by dholmes-ora** [comment]
Thanks @tstuefe !

Can I get second review please.

**Review by robehn — APPROVED** [review]
Thanks!

**Comment by dholmes-ora** [comment]
Thanks @robehn !

/integrate

**Pull request closed without merging**
